로그인,프로필 화면 [STEP 2] liam, wily

liam-777 opened the pull request:
@wonhee009 항상 상세한 피드백 감사드립니다🚀 STEP2 구현 내용 PR 요청드립니다!
 
# 주요 구현 내용
* `LabeldImageView` : `UIStackView`
  * 하단에 아이콘(`UIImageView`)과 레이블(`UILabel`)을 묶어 커스텀 UIStackView를 선언했습니다.  

* `ProfileView` : `LabeledImageView`
  * 가운데 프로필 화면의 구조가 저희가 선언해둔 `LabeledImageView`와 동일하다고 생각되어 이를 상속받는 `ProvileView`를 선언했습니다.
  * spacing, corener radius, label font, imageview 제약 조건을 추가

* `ProfileViewController`
  * 컴포넌트들을 모두 선언해 놓고 `configure컴포넌트이름` 각각에서 오토레이아웃 제약 조건을 부여하였습니다.
  
# 고민했던 점 및 궁금했던 점

1. 최대한 Constant 사용을 자제하려고 하였으나 오토레이아웃을 쓰다보니 Constant 값들을 쓰게 되었습니다.(예시: 이미지 크기, 아이콘 크기, View들간의 간격) 화면이 작은 아이팟 터치 7세대와 아이폰 13 프로의 경우 결과물의 차이가 약간씩 나게 되는데 이를 어떤식으로 처리하는 것이 좋은지 궁금합니다.

2. ProfileViewController.swift에서 Constants 라는 Struct를 만들어 상수들을 관리하려고 하는데 모든 상수들을 이 구조체에 넣어 관리하려 하다가 의미가 있는(그냥 사용하면 의미가 애매모호한) 상수들만 넣자라는 방식을 채택하려고 합니다. UI구현을 사용하는 상수들을 Struct에 넣어서 함께 관리하는 것이 좋을까요?
(추가적으로 상수들을 관리할때 Struct 대신 Enum을 쓰면 협업시 실수로 인스턴스를 생성하는 것을 방지할 수 있다고 하여 좋다고 들었던 기억이 있습니다. 실제로 Enum을 사용하기도 하는지 궁금합니다.)

3. ProfileView같은 경우는 LabeledImageView를 상속받았습니다. 두개의 뷰는 유사하고 차이점은 ProfileView의 경우 텍스트 타입과 이미지 크기 그리고 코너가 둥글어야한다는 점이었습니다. 
따라서 상속을 받고 다른 부분을 추가적으로 구현하였는데 이러한 방식이 맞는지 궁금합니다.

4. 문제에서 화면 크기의 30%의 corner radius라는 조건이 있어 ProfileView에서 layoutsubviews()에서 corner radius를 설정하도록 하였습니다. 이 방법 말고 더 나은 방법이 있는지 궁굼합니다.

5. 저희가 LabeldImageView에 UILabel, UIImageView를 프로퍼티로 갖고 있는데 이를 private으로 선언하고 외부에서 메서드로 접근할 수 있도록 설계하는 것이 맞는 방법인가요? 
 
- 상속을 받게되는 경우는 자바나 다른 언어와 같이 상속 클래스에게 허용하는 접근 제한자가 없는데, 이 경우에도 메서드로 접근해야만 하는 것인지?
- private으로 프로퍼티를 선언하면 상속받은 클래스에서는 오버리이딩 조차 하지 못하게 되는데 오버라이딩이 필요할 때는 불가피하게 private을 제거해야 하는 것인지?

+ Apple의 여러 UIKit의 사례들을 보면 프로퍼티에 직접 접근할 수 있도록 설계된 경우가 많았는데 어떤 방법이 더 나은 방법인지 궁굼합니다.

liam-777 committed:
feat: step2 프로필 화면 구현

wilybear committed:
feat: step2 프로필 뷰 및 하단 스택뷰 구현

- ProfileView 구현
- 하단 스텍뷰 구현
- Step1 PR 반영
- 리펙토링

wonhee009 commented on LoginProfile/ViewController/ProfileViewController.swift:
```diff
@@ -0,0 +1,150 @@
+        stackView.distribution = .fillEqually
+        stackView.alignment = .center
+        return stackView
+    }()
```
이 아이들만 `lazy`로 선언한 이유가 있을까요?🤔

liam-777 replied:
let으로 모두 변경해서 반영했습니다!

wonhee009 commented on LoginProfile/ViewController/ProfileViewController.swift:
```diff
@@ -0,0 +1,150 @@
+    private lazy var bottomLabeledImageStackView: UIStackView = {
+        let stackView: UIStackView = UIStackView(arrangedSubviews: [chatLabeledImageView,
+                                                                    phoneLabeledImageView,
+                                                                    storyLabeledImageView])
```
`arrangedSubviews`를 사용해보셨군용👍

wonhee009 replied:
스택뷰의 subview들을 생성 시점부터 넣어주는 이유가 있을까요?🤔

wonhee009 commented on LoginProfile/ViewController/ProfileViewController.swift:
```diff
@@ -0,0 +1,150 @@
+    }
+}
+
+extension ProfileViewController {
```
`viewDidLoad()`를 제외한 모든 메서드가 extension으로 정의되어 있는데 따로 이유가 있을까요?🤔

wonhee009 commented on LoginProfile/ViewController/ProfileViewController.swift:
```diff
@@ -0,0 +1,150 @@
+        configureGiftButton()
+        configureWonButton()
+        configureDivider()
+        configureProfileView()
```
아무래도 하나의 화면에 들어가는 UI 요소가 많다보니 `configure+UI요소`라는 메서드가 많이 생기네요.
현재 화면보다 많은 UI 요소가 존재한다면 무한정으로 configure메서드가 생길거 같은데요.
UI요소별로 메서드를 나누기보다는 역할에 초점을 맞춰 보는건 어떨까요?
```swift
func setupSubviews() {
     view.addSubview(profileView)
     ...
}

func setupConstraints() { ... {
```

이런식으로요☺️

wonhee009 commented on LoginProfile/View/LabeledImageView.swift:
```diff
@@ -0,0 +1,56 @@
+    
+    func setDescriptionLabelText(_ text: String?) {
+        imageDescriptionLabel.text = text
+    }
```
해당 요소에서 image나 text는 처음 결정되면 변경될 일이 없는거로 이해했는데요.
init시점에 image나 text를 주입해주는건 어떨까요?
```swift
init(image: UIImage?, text: String?) {
     ...
}
```

wonhee009 reviewed (changes requested):
@liam-777 @wilybear step2 수고하셨습니다!🎉
코멘트 남겨두었으니 확인해주세요🙏

#### 1

이 경우는 화면마다 다를거 같습니다.
어떤 화면은 constant하게 간격을 가져가는 경우도 있고 아닌 경우도 있어서요.
보통 이미지는 전 기종 같은 크기로 가져가거나 아이폰/아이패드 이렇게 나눠서 두벌로 가져가는 경우도 있고요.

#### 2.

저는 상수들을 하나로 관리하는걸 좀 더 선호하는데요. (특히 디자인 요소)
하나로 관리하면 나중에 디자인만 수정할때 편하더라구요.
넵 enum 사용하고 있습니다.

#### 3. 

profileView와 LabeldImageView 사이에 다른 점이 corner 뿐이라면 상속보다는 init 시점에 둥글게 할지 안할지만 Bool 값으로 주입 받는 방식으로도 해결할 수 있을거 같습니다.
(init에 관해서는 해당 코드에도 코멘트를 달아두었습니다:))

#### 4. 

configure 내에서도 가능할거 같습니다.

#### 5.

저는 최대한 적은 요소들이 수정에 열려있어야 한다고 생각해요.
따라서 메서드로 접근하도록 하는게 좀 더 맞지 않을까 생각하기도 하구요.
다만, 현재 구조상 화면의 요소들이 바뀔 일이 없어서 모두 private으로 닫고 init 시점에 주입하는 방식이 좀 더 낫지 않나 생각합니다.

liam-777 committed:
feat: step2 PR 리뷰 반영

liam-777 committed:
feat: step2: PR 반영, Profileview 삭제

liam-777 commented:
PR 리뷰 남겨주신 부분 반영했습니다!
커스텀 뷰 프로퍼티들은 private으로 관리시키고 생성자를 통해서 기존 ProfileView까지 대응할 수 있도록 하나로 만들었습니다.
많이 해맸던 부분인데 좋은 방향, 좋은 피드백 남겨 주셔서 감사합니다 👏

wonhee009 commented on LoginProfile/ViewController/ProfileViewController.swift:
```diff
@@ -0,0 +1,152 @@
+    }()
+    
+    private let storyLabeledImageView: LabeledImageView = {
+        
```
```suggestion
```

wonhee009 commented on LoginProfile/ViewController/ProfileViewController.swift:
```diff
@@ -0,0 +1,152 @@
+    }
+    
+    func setupConstraints() {
+        profileView.translatesAutoresizingMaskIntoConstraints = false
```
이 부분들은 UI 요소 생성 시점에 지정해줘도 될거 같은데요.🤔

wonhee009 commented on LoginProfile/ViewController/ProfileViewController.swift:
```diff
@@ -0,0 +1,152 @@
+        setupConstraints()
+    }
+    
+    func setupSubviews() {
+        view.addSubview(profileView)
+        view.addSubview(closeButton)
+        view.addSubview(topButtonStackView)
+        view.addSubview(bottomLabeledImageStackView)
+        view.addSubview(dividerView)
+    }
+    
+    func setupConstraints() {
```
이런 UI 설정을 위한 메서드들은 외부에서 접근할 가능성이 있는걸까요?🤔

wonhee009 commented on LoginProfile/ViewController/ProfileViewController.swift:
```diff
@@ -0,0 +1,152 @@
+        topButtonStackView.addArrangedSubview(favouriteButton)
+        
+        let topButtonCount = CGFloat(topButtonStackView.arrangedSubviews.count)
+        let topButtonStackViewWidth: CGFloat = Constants.topButtonHeight * topButtonCount + Constants.topButtonSpacing * (topButtonCount - 1)
```
상단의 버튼들을 담고 있는 stackView의 width를 지정해주었네요!
혹시 autoLayout에서 `greaterThanOrEqual`이나 `lessThanOrEqual`에 대해서 알고 계실까요?
해당 constraint relation을 사용한다면 width를 계산하지 않고 `x` 버튼과의 관계만으로도 화면을 구성할 수 있을거 같습니다!

wonhee009 commented on LoginProfile/ViewController/ProfileViewController.swift:
```diff
@@ -0,0 +1,152 @@
+    }
+    
+    enum Constants {
+        static let topButtonHeight: CGFloat = 26
+        static let topButtonSpacing: CGFloat = 12
```
`CGFloat`로 타입을 명시해주셨군용👍

wonhee009 commented on LoginProfile/View/LabeledImageView.swift:
```diff
@@ -0,0 +1,76 @@
+    
+    private var isRoundCorner: Bool = false
+    
+    override private init(frame: CGRect) {
+        super.init(frame: frame)
+        axis = .vertical
+        spacing = 15
+        distribution = .fill
+        alignment = .center
+        configureUI()
+    }
+    
+    convenience init(image: UIImage?, text: String?, isRoundCorner: Bool, textFont: UIFont? = .preferredFont(forTextStyle: .subheadline)) {
+        self.init(frame: .zero)
+        imageView.image = image
+        imageDescriptionLabel.text = text
+        self.imageDescriptionLabel.font = textFont
+        self.isRoundCorner = isRoundCorner
+    }
+    
+    convenience init(image: UIImage?, text: String?) {
+        self.init(image: image, text: text, isRoundCorner: false)
+    }
+
```
`init` 메서드가 많은데요.
```swift
init(image: UIImage?, text: String?, isRoundCorner: Bool = false) {
     super.init(frame: .zero)
     configureUI()
     if let image = image {
          setImage(image)
     }
     if let text = text {
          setText(text)
     }
}

private func configureStackView() {
     axis = .vertical
     spacing = 15
     ...
}

private func setImage(_ image: UIImage) {
     ...
}

private func setText(_ text: String) {
     ...
}
```

이렇게는 불가능한가요?.?
textFont도 하나의 폰트만 사용할거 같은데 init 시에 주입 받는 이유가 따로 있을까요?

liam-777 replied:
가운데 프로파일에 사용할 폰트와 하단 부분과는 폰트가 달라서 주입했습니다.

혹시  textFont는 `func setTextFont(_ font: UIFont?)` 와 같이 `LabeledImageView` 에서 선언해두고

`ProfileViewController` 쪽에서
```swift
func configureProfileView() { 
    ...
    profileView.setTextFont(.title3)
}
```
와 같이 설정하는 것이 더 나을까요?

wonhee009 replied:
제가 이해한 바로는
- profileView: corner 둥글게 처리, font1
- 하단뷰: corner 처리 x, font2

로 이해했습니다.

그렇다면 
```swift
class LabeledImageView {
     enum ViewType {
          case profile
          case normal
     }
    
     init(type: LabeledImageView = .normal, image: UIImage?, text: String?) {
          ...
     }
}
```

이렇게 타입을 나눈 후 주입 받는건 어떨까요?

liam-777 replied:
와우 .. 👍👍 감사합니다. 반영해보겠습니다!

wonhee009 reviewed (changes requested):
수고하셨습니다!
몇몇 코멘트 남겨두었으니 확인 부탁드립니다!🙏

liam-777 committed:
feat: step2 PR 반영

- LabeldImageView init 수정
- ProfileViewController private 접근제한자 지정
- topButtonStackView autolayout 수정

wonhee009 commented on LoginProfile/View/LabeledImageView.swift:
```diff
@@ -53,24 +49,35 @@ class LabeledImageView: UIStackView {
+        
+        if viewType == .profile {
             let radius: CGFloat = bounds.size.width * 0.3
             imageView.layer.cornerRadius = radius
         }
     }
     
+    private func configureStackViewProperties() {
+        axis = .vertical
+        spacing = viewType == .normal ? 15 : 5
+        distribution = .fill
+        alignment = .center
+    }
+    
     private func configureUI(){
         addArrangedSubview(imageView)
         addArrangedSubview(imageDescriptionLabel)
         
-        imageView.clipsToBounds = isRoundCorner
-    }
-    
-    func setupImageViewWidthEqaulToLabeledImageView() {
-        imageView.widthAnchor.constraint(equalTo: widthAnchor).isActive = true
+        imageView.clipsToBounds = viewType == .profile
+        
+        let font: UIFont = viewType == .normal ? .preferredFont(forTextStyle: .subheadline) : .preferredFont(forTextStyle: .title3)
+        imageDescriptionLabel.font = font
+        
     }
```
private func으로 `configureProfileUI()`, `configureNormalUI()`로 분리해도 좋을거 같습니다!
```swift
private func configureUI() {
     switch viewType {
          case .profile:
               configureProfileUI()
     }
}
```
이런 식으로 분리를 한다면 이후에 프로필에 대한 디자인 스펙이 변경된다면 해당 메서드만 살펴보면 좀 더 유지보수하기 편하지 않을까 생각이 들어서요.
현재는 삼항 연산자를 사용해서 관리하고 있는데 삼항 연산자 성능에 대해서도 알아보시면 좋을거 같습니다.☺️
또한, 삼항 연산자가 가독성을 해치는 경우가 많아 저는 사용을 지양하기도 합니다!

liam-777 replied:
항상 구조적으로 꼼꼼하게 봐주셔서 너무 감사합니다 👏 덕분에 이번 프로젝트 동안 많이 배워간 것 같습니다!

wonhee009 reviewed (approved):
@liam-777 @wilybear 이번 프로젝트도 수고 많으셨습니다!🎉
코드로 UI를 작성하는게 처음였다면 많이 혼란스러우셨을거 같아요. 스토리보드나 xib와 다르게 눈에 직관적으로 보이지 않고, 빌드를 해봐야만 결과물이 보이기 때문에 많은 어려움도 있었을거 같구요.
그래도 코드를 작성하기 전에 화면의 요소들을 어떻게 구성할지 생각해보면서 작성하는 것도 하나의 재미인거 같습니다!☺️
이번 프로젝트는 여기서 마무리해도 좋을거 같습니다.
코멘트를 하나 남겨두긴 했는데 확인해보시고 디벨롭해보셔도 좋을거 같습니다!
수고하셨습니다!👍
